### PR TITLE
New package: launcher

### PR DESCRIPTION
**Pull request by supechicken**
Changes from #4739 
- Use websocket
- Added some new features

Tested on
- Cloudready v85
- Chrome OS v87

### CODE REVIEW

File: crew_integration/main.sh [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+YELLOW='\e[1;33m';
+RESET='\e[0m'
+CREW_PREFIX=/usr/local
+CREW_LIB_PREFIX=${CREW_PREFIX}/lib
+EXTENSION_PREFIX=~/MyFiles/.extension
+PWA_PREFIX=${CREW_LIB_PREFIX}/pwa
+SERVER=${PWA_PREFIX}/server.rb
+SENDER=${PWA_PREFIX}/send.rb
+app_path=${EXTENSION_PREFIX}/apps/${2}
+extension_id=id_id
```
Comment (uberhacker) [inline crew_integration/main.sh:14]: Should we use environment variables if they are set?  For example:
```
CREW_PREFIX="${CREW_PREFIX:-/usr/local}"
```

File: packages/crew_integration.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+     FileUtils.rm_rf "#{CREW_PREFIX}/lib/pwa/"
+     FileUtils.rm_rf "#{HOME}/MyFiles/.extension"
+  end
+    
+  case ARGV[0]
+  when 'reinstall'
+     puts "Since you are reinstalling this package, please run 'pkill ruby' if installation failed".lightblue
+  when 'upgrade'
+     puts "Since you are upgrading this package, please run 'pkill ruby' if installation failed".lightblue
+  end
```
Comment (uberhacker) [inline packages/crew_integration.rb:23]: Why not just run `pkill ruby` as part of the process?

File: packages/crew_integration.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      FileUtils.ln_s "#{CREW_PREFIX}/bin/crew_integration", "#{CREW_DEST_PREFIX}/bin/pwashortcut"
+      FileUtils.mkdir_p "#{CREW_DEST_HOME}/MyFiles/.extension/"
+      FileUtils.mv Dir.glob('extension/*'), "#{CREW_DEST_HOME}/MyFiles/.extension/"
+      FileUtils.mv Dir.glob('*'), "#{CREW_DEST_PREFIX}/lib/pwa/"
+      FileUtils.ln_s "#{CREW_PREFIX}/lib/pwa/send.rb", "#{CREW_DEST_PREFIX}/lib/pwa/sender.rb"
+    end 
+  end
+  def self.postinstall
+      case ARGV[0]
+      when 'reinstall', 'install'
```
Comment (uberhacker) [inline packages/crew_integration.rb:45]: Why is this necessary?  If you run `crew postinstall crew_integration`, you won't see this message?

**Comment by uberhacker** [comment]
@supechicken: Wow, it looks like you put a lot of effort into this package!  I will test it out and provide some feedback.

File: packages/crew_integration.rb (hunk 2 above)
Comment (supechicken) [reply packages/crew_integration.rb:23]: Isn't `crew` based on ruby lol

Comment (uberhacker) [reply packages/crew_integration.rb:23]: Not as a service though, right?  I was thinking something like:
```
  case ARGV[0]
  when 'reinstall', 'install'
     system 'pkill ruby'
     FileUtils.rm_rf "#{CREW_PREFIX}/lib/pwa/"
     FileUtils.rm_rf "#{HOME}/MyFiles/.extension"
  end
```

Comment (supechicken) [reply packages/crew_integration.rb:23]: The installation will stop immediately after `pkill`, I am finding a better solution on closing the websocket connection now:)

File: packages/crew_integration.rb (hunk 3 above)
Comment (supechicken) [reply packages/crew_integration.rb]: We don't need this part at upgrade as it doesn't need to re-complete the setup again, added `postinstall` now

File: crew_integration/main.sh (hunk 1 above)
Comment (supechicken) [reply crew_integration/main.sh:14]: Done

**Comment by uberhacker** [comment]
What about calling the package `pwa_integration` or simply `pwa`?

**Comment by supechicken** [comment]
@uberhacker I think the package name is too long also, how about `integration`?

File: packages/crew_integration.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      FileUtils.ln_s "#{CREW_PREFIX}/bin/crew_integration", "#{CREW_DEST_PREFIX}/bin/pwashortcut"
+      FileUtils.mkdir_p "#{CREW_DEST_HOME}/MyFiles/.extension/"
+      FileUtils.mv Dir.glob('extension/*'), "#{CREW_DEST_HOME}/MyFiles/.extension/"
+      FileUtils.mv Dir.glob('*'), "#{CREW_DEST_PREFIX}/lib/pwa/"
+      FileUtils.ln_s "#{CREW_PREFIX}/lib/pwa/send.rb", "#{CREW_DEST_PREFIX}/lib/pwa/sender.rb"
+    end 
+  end
+  def self.postinstall
+      case ARGV[0]
+      when 'reinstall', 'install', 'upgrade', 'postinstall'
```
Comment (uberhacker) [inline packages/crew_integration.rb:45]: Why include this logic at all?

File: packages/crew_integration.rb [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  version '2.0'
+  source_url 'file:///dev/null'
+  source_sha256 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'
+  
+  depends_on "graphicsmagick"
+    
+  case ARGV[0]
+  when 'reinstall', 'install'
+     FileUtils.rm_rf "#{CREW_PREFIX}/lib/pwa/"
+     FileUtils.rm_rf "#{HOME}/MyFiles/.extension"
```
Comment (uberhacker) [inline packages/crew_integration.rb:15]: Should these lines be in the `def self.remove` section?

File: packages/crew_integration.rb (hunk 4 above)
Comment (supechicken) [reply packages/crew_integration.rb:45]: We don't need this part at upgrade as it doesn't need to re-complete the setup again, might be better using `unless`

File: packages/crew_integration.rb (hunk 5 above)
Comment (supechicken) [reply packages/crew_integration.rb]: Moved to `remove` section

**Comment by daemonspudguy** [comment]
Can @uberhacker merge this now?

**Comment by uberhacker** [comment]
@daemonspudguy: I wasn't able to get this to work.  There have been a few more commits since I checked last so I can give it another try.  You are welcome to test also.

**Comment by supechicken** [comment]
Update v3.0

- Commands now will pass directly through URL instead of separate htmls
- Move all app icons from `PWA_PREFIX` into the extension
- Fix bugs

**Comment by daemonspudguy** [comment]
In order for the gems to install, you need to make two symlinks. run `ln -s /bin/mkdir /usr/local/bin/mkdir && ln -s /usr/bin/install /usr/local/bin/install`.

**Comment by supechicken** [comment]
@daemonspudguy It works for me without symlink. By the way thanks for reminding me, change made

File: packages/integration.rb [hunk 6]
```diff
@@ -27,12 +27,12 @@ def self.install
     system "gem install em-websocket"
     FileUtils.safe_unlink "#{CREW_PREFIX}/bin/mkdir"
     FileUtils.safe_unlink "#{CREW_PREFIX}/bin/install"
```
Comment (uberhacker) [inline packages/integration.rb:29]: I don't think this is a good idea.  We shouldn't be removing files from other packages.

Comment (supechicken) [reply packages/integration.rb]: Removed

**Comment by daemonspudguy** [comment]
I've been using it and it works a treat.

File: packages/integration.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
+require 'package'
+
+class Integration < Package
+  description 'Improves integration with Chrome OS shell and Chromebrew GUI applications.'
+  homepage 'https://github.com/skycocker/chromebrew/crew_integration'
+  version '3.0'
```
Comment (uberhacker) [inline packages/integration.rb:6]: Missing `compatibility 'all'`?

File: packages/integration.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  case ARGV[0]
+  when 'reinstall'
+     puts "Since you are reinstalling this package, please run 'pkill ruby' if installation failed".lightblue
+  when 'upgrade'
+     puts "Since you are upgrading this package, please run 'pkill ruby' if installation failed".lightblue
+  end
+
+  def self.install
+    # system "curl https://codeload.github.com/skycocker/chromebrew/zip/master -#o crew_integration.zip"
+    system "curl https://codeload.github.com/supechicken/chromebrew/zip/supechicken-patch-1 -#o crew_integration.zip"
```
Comment (uberhacker) [inline packages/integration.rb:21]: Can you please rename your branch to something more descriptive like `pwa-integration` or something similar?  Also, you should be tagging your branch with a release (version).

File: packages/integration.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    puts 'Archive downloaded'.lightgreen
+    system "unzip crew_integration.zip > /dev/null"
```
Comment (uberhacker) [inline packages/integration.rb:23]: No need for `> /dev/null`.  It's fine (and better for debugging, if necessary) to show the output.

File: packages/integration.rb [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    FileUtils.mkdir_p "#{CREW_DEST_PREFIX}/bin"
+    FileUtils.ln_s '/bin/mkdir', "#{CREW_PREFIX}/bin/mkdir" unless File.exist? "#{CREW_PREFIX}/bin/mkdir"
+    FileUtils.ln_s '/usr/bin/install', "#{CREW_PREFIX}/bin/install" unless File.exist? "#{CREW_PREFIX}/bin/install"
```
Comment (uberhacker) [inline packages/integration.rb:26]: These symlinks should not be needed.  My guess is your `PATH` environment is missing `/bin` and `/usr/bin` which should be already in the default path.

File: packages/integration.rb [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    system "gem install em-websocket"
+    FileUtils.chmod_R '+rwrwrw', './'
```
Comment (uberhacker) [inline packages/integration.rb:28]: I'm confused why this is needed.

File: integration/extension/main.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
+var ws = new WebSocket('ws://localhost:25500','protocol');
+var opt = new URLSearchParams(location.search);
+var cmd = opt.get('cmd');
+var name = opt.get('friendly_name')
+document.title = name;
+document.querySelector("link[rel~='icon']").href = `/icon/${cmd}.png`;
+command(cmd)
+function command(cmd) {
```
Comment (uberhacker) [inline integration/extension/main.js:8]: Shouldn't `command(cmd)` be called after defining the function?

File: integration/main.sh [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+YELLOW='\e[1;33m';
+RESET='\e[0m'
+CREW_PREFIX="${CREW_PREFIX:-/usr/local}"
+CREW_LIB_PREFIX=${CREW_PREFIX}/lib
+EXTENSION_PREFIX=${HOME}/MyFiles/.extension
+PWA_PREFIX=${CREW_LIB_PREFIX}/pwa
+SERVER=${PWA_PREFIX}/server.rb
+SENDER=${PWA_PREFIX}/send.rb
+appicon_path=${EXTENSION_PREFIX}/icon
+extension_id=id_id
```
Comment (uberhacker) [inline integration/main.sh:14]: Use uppercase variables?

File: packages/integration.rb [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+         EM::WebSocket.run(:host => "0.0.0.0", :port => 25500) do |ws|
+           ws.onopen { |handshake| }
+           ws.onmessage { |message|
+             `sed -i "s/id_id/#{message}/g" #{CREW_PREFIX}/bin/crew_integration`
+             EM::WebSocket.stop
+           }
+         end
+        }
+      end
+      puts "This package contain a terminal shortcut, execute the following to install".lightblue
```
Comment (uberhacker) [inline packages/integration.rb:64]: `puts "This package contains a terminal shortcut.  To install, execute the following:".lightblue` ?

File: packages/integration.rb [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+      puts "crew_integration -i".lightblue
+      puts
+      puts "Run 'crew_integration -h' for more usage of this package".lightblue
```
Comment (uberhacker) [inline packages/integration.rb:67]: `puts "Run 'crew_integration -h' for help.".lightblue` ?

File: packages/integration.rb [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+      FileUtils.mv Dir.glob('extension/*'), "#{CREW_DEST_HOME}/MyFiles/.extension/"
+      FileUtils.mv Dir.glob('*'), "#{CREW_DEST_PREFIX}/lib/pwa/"
+      FileUtils.ln_s "#{CREW_PREFIX}/lib/pwa/send.rb", "#{CREW_DEST_PREFIX}/lib/pwa/sender.rb"
+    end 
+  end
+  def self.postinstall
+      unless ARGV[0] == "upgrade"
+        require 'em-websocket'
+        puts 
+        puts "To complete the installation, install the extension by following:".lightgreen
```
Comment (uberhacker) [inline packages/integration.rb:46]: `puts "To complete the installation, execute the following:".lightblue` ?

File: packages/integration.rb [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+        puts "Go to chrome://extensions/".lightgreen
+        puts "Switch on Developer Mode".lightgreen
+        puts "Click 'Load Unpacked'".lightgreen
+        puts "Click the 'More' button (icon with three dots) in top-right side of the new window".lightgreen
```
Comment (uberhacker) [inline packages/integration.rb:50]: `puts "Click the 'More' button (icon with three dots) in the top right corner of the new window".lightgreen` ?

Comment (supechicken) [reply packages/integration.rb]: Done

File: packages/integration.rb (hunk 7 above)
Comment (supechicken) [reply packages/integration.rb:6]: Done

File: packages/integration.rb (hunk 8 above)
Comment (supechicken) [reply packages/integration.rb:22]: I think it is better to keep all stuff inside this repo, can I open a new branch here?

File: packages/integration.rb (hunk 11 above)
Comment (supechicken) [reply packages/integration.rb:29]: In my system, the permissions are `0000` after unzipping the archive

File: packages/integration.rb (hunk 9 above)
Comment (supechicken) [reply packages/integration.rb]: Removed

File: integration/main.sh (hunk 13 above)
Comment (supechicken) [reply integration/main.sh]: Done

**Comment by supechicken** [comment]
> I've been using it and it works a treat.

@daemonspudguy Thank you for testing it out :)

File: packages/integration.rb (hunk 14 above)
Comment (supechicken) [reply packages/integration.rb]: Done

File: integration/extension/main.js (hunk 12 above)
Comment (uberhacker) [reply integration/extension/main.js:8]: Isn't that only true after encapsulating the onready event?

File: integration/main.sh [hunk 18]
```diff
@@ -33,34 +33,35 @@ case ${1} in
        if [[ ${DESKTOP_FILE} != '' ]]; then
```
Comment (uberhacker) [inline integration/main.sh:33]: Exit with usage message if `${2}` is missing?
`DESKTOP_FILE=$( find ${CREW_PREFIX} -name ${CMD}.desktop | sed -n '1p' )`?

Comment (uberhacker) [reply integration/main.sh:33]: Easier?  `if [ -n ${DESKTOP_FILE} ]; then`

File: packages/integration.rb [hunk 19]
```diff
@@ -41,6 +41,11 @@ def self.install
     end 
   end
   def self.postinstall
+      @cmd = 'nohup crew_integration -s'
+      @check = `grep -c "#{@cmd}" #{HOME}/.bashrc`
```
Comment (uberhacker) [inline packages/integration.rb:45]: Use `$(...)` instead of backticks?

**Comment by supechicken** [comment]
Update v5.0

- Use Chrome Apps instead of PWA to make the shortcut more native experence
- Icon of the command will find on local now
- Use Ruby

**Comment by supechicken** [comment]
Screenshot 2021-02-11 at 2 46 43 PM

File: packages/crew_profile.rb [hunk 20]
```diff
@@ -3,17 +3,17 @@
...
 
   depends_on "xdg_base"
 
   def self.install
-    system "mkdir", "-p", "#{CREW_DEST_PREFIX}/etc"
-    system "mkdir", "-p", "#{CREW_DEST_HOME}"
-    system "ln", "-s", "#{ENV['HOME']}/.profile", "#{CREW_DEST_PREFIX}/etc/profile"
+    FileUtils.mkdir_p "#{CREW_DEST_PREFIX}/etc"
+    FileUtils.mkdir_p CREW_DEST_HOME
+    FileUtils.ln_s "#{Dir.home}/.profile", "#{CREW_DEST_PREFIX}/etc/profile"
```
Comment (uberhacker) [inline packages/crew_profile.rb:16]: This can be `"#{HOME}/.profile"`.

**Comment by uberhacker** [comment]
I'll have to test this out.  It looks promising!

File: packages/crew_profile.rb (hunk 20 above)
Comment (supechicken) [reply packages/crew_profile.rb]: Change made

File: packages/launcher.rb [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+  homepage 'https://github.com/supechicken/chromebrew/tree/master/launcher'
+  version '5.0'
+  source_url 'https://github.com/supechicken/chromebrew/raw/master/launcher/launcher.rb'
+  source_sha256 '117615ee5f2cf5096c8fef274cfa3cc94c2161b5b23e080ff199081c18b5c0a7'
+  
+  depends_on 'libyaml'
+  depends_on 'imagemagick7'
+
+  def self.install
+    system('gem', 'install', 'em-websocket')
```
Comment (uberhacker) [inline packages/launcher.rb:14]: Can we include the gem dependency files in the package?  I'm thinking this could be done similar to how we add perl or python dependencies in other packages.  It might even make sense to package the `em-websocket` gem in a separate package and add that as a dependency here.

File: launcher/screenshot/README.md [hunk 22]
```diff
@@ -0,0 +1 @@
+
```
Comment (uberhacker) [inline launcher/screenshot/README.md:1]: Empty file?

File: launcher/launcher.rb [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+          system(message + '&')
+        }
+      end
+    }
+  end
+end
+    
+case ARGV[0]
+when 'start'
+  start_server
```
Comment (uberhacker) [inline launcher/launcher.rb:138]: Should we have some feedback to confirm the server has started?

File: launcher/launcher.rb [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+when 'new'
+  new ARGV[1]
```
Comment (uberhacker) [inline launcher/launcher.rb:140]: Will this return a clear message if `ARGV[1]` is missing?  Would it make sense to echo the usage text instead?

File: launcher/launcher.rb [hunk 25]
```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env ruby
+CREW = `dirname $(which crew)`
+
+# use chromebrew variable
+require "#{CREW}/../lib/crew/lib/const"
+require "#{CREW}/../lib/crew/lib/color"
```
Comment (uberhacker) [inline launcher/launcher.rb:6]: No need to create a `CREW` variable?
```
require "#{CREW_PREFIX}/lib/crew/lib/const"
require "#{CREW_PREFIX}/lib/crew/lib/color"
```

File: launcher/launcher.rb [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+require 'em-websocket'
+require 'fileutils'
+
+PID = CREW_PREFIX + '/tmp/server.pid'
+LOG = CREW_PREFIX + '/tmp/server.log'
+PWA_PREFIX = CREW_PREFIX + '/lib/pwa'
+PORT = 25500
+PATH = "#{CREW_PREFIX}/share/pixmaps " + "/usr/local/share/icons " + "#{Dir.home}/.local/ " + PWA_PREFIX
```
Comment (uberhacker) [inline launcher/launcher.rb:14]: You could use `#{HOME}` instead of `#{Dir.home}`.

File: launcher/launcher.rb [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+    Process.kill 'KILL', File.read(PID).to_i if File.exist?(PID)
+  rescue
+  end
+end
+
+
+def desktop_icon_name_finder (app)
+  # find out the info, name and icon of the command
+    
+  @desktop = `find / -name #{app}.desktop 2> /dev/null | sed -e '/stateful_partition/d' | tail -n1`.tr("\n", '')
```
Comment (uberhacker) [inline launcher/launcher.rb:69]: Is there a consistent pattern for where `#{app}.desktop` is located?  If so, why would we need to `find`?

File: launcher/launcher.rb [hunk 28]
```diff
@@ -0,0 +1,147 @@
...
+    
+  # abort if a desktop file not found for the command
+  if @desktop == '' then
```
Comment (uberhacker) [inline launcher/launcher.rb:72]: Could you use `unless File.exists? "/path/to/#{app}.desktop"` instead?

File: launcher/launcher.rb [hunk 29]
```diff
@@ -0,0 +1,147 @@
...
+    puts "Cannot find the .desktop file for #{app}, using default Chromebrew icon... 
+(If you want to use customize icon, put the icon file to #{CREW_PREFIX}/lib/pwa/ and name it to '#{app}.png')".yellow
+    FileUtils.mkdir_p PWA_PREFIX
+    system('curl', '-L#', 'https://github.com/supechicken/chromebrew/raw/master/images/brew546.png', '-o', PWA_PREFIX + '/brew.png')
```
Comment (uberhacker) [inline launcher/launcher.rb:76]: Should we replace `supechicken` with `skycocker`?

File: launcher/launcher.rb [hunk 30]
```diff
@@ -0,0 +1,147 @@
...
+    @icon = PWA_PREFIX + '/brew.png'
+    @icon = PWA_PREFIX + "/#{app}.png" if File.exist?(PWA_PREFIX + "/#{app}.png")
```
Comment (uberhacker) [inline launcher/launcher.rb:78]: I see no use for line 77.  You overwrite on line 78?

File: launcher/launcher.rb [hunk 31]
```diff
@@ -0,0 +1,147 @@
...
+    @name = app.capitalize
+  else
+    @icon = File.readlines(@desktop).select { |line| line =~ /^Icon/ }[0].partition('=')[2].tr("\n", '')
+    @icon = `find #{PATH} -iname #{@icon}.png -o -iname #{@icon}.xpm 2> /dev/null | sed -e '/stateful_partition/d' |\
+             sed -e 's:/usr/local/share/icons/hicolor/::g' | sort -n | sed -e 's:^:/usr/local/share/icons/hicolor/:g' |\
+             tail -n1`.tr("\n", '')
```
Comment (uberhacker) [inline launcher/launcher.rb:84]: Don't we already know the path?  Why do we need to find?

File: launcher/launcher.rb [hunk 32]
```diff
@@ -0,0 +1,147 @@
...
+
+    @name = File.readlines(@desktop).select { |line| line =~ /^Name/ }[0].partition('=')[2].tr("\n", '')
+    @info = File.readlines(@desktop).select { |line| line =~ /^Comment/ }[0].partition('=')[2].tr("\n", '')
+  end
+  @info = `crew search -d #{app} | head -1` unless @info
+end
+
+def new (app)
+  # abort when command not found
+  abort "#{app}: command not found".lightred unless system("which #{app} > /dev/null 2>1")
```
Comment (uberhacker) [inline launcher/launcher.rb:95]: This will output `: command not found` if `app` is missing.  Not a helpful message to the user.

File: launcher/README.md [hunk 33]
```diff
@@ -0,0 +1,22 @@
...
+- Chrome/Chromium OS based system (CloudReady) with 56+ version
+- Stable channel (recommend)
+
+## Installation
+- This package is available at `crew`, install it by execute the following:
+```bash
+crew install launcher
+```
+## Help needed?
+- Submit an new issues [here](https://github.com/skycocker/chromebrew/issues)<sup>*</sup>
```
Comment (uberhacker) [inline launcher/README.md:16]: `Submit new issues`?  No need for `an`.

File: launcher/launcher.rb [hunk 34]
```diff
@@ -0,0 +1,147 @@
...
+
+  desktop_icon_name_finder(app)
+  
+  @working_on_dir = "#{Dir.home}/MyFiles/#{app}.launcher/"
```
Comment (uberhacker) [inline launcher/launcher.rb:99]: `#{HOME}` instead of `#{Dir.home}`?

Comment (supechicken) [reply launcher/launcher.rb]: Done

File: launcher/launcher.rb (hunk 24 above)
Comment (supechicken) [reply launcher/launcher.rb]: Added

File: launcher/launcher.rb (hunk 32 above)
Comment (supechicken) [reply launcher/launcher.rb]: Changed to "#{app}: application not installed"

**Comment by supechicken** [comment]
Just added a part in `/launcher/launcher.rb` to create a daemon for the WebSocket server (to replace `nohup`), now the server will not close unless `kill` :)

**Comment by supechicken** [comment]
I changed the files a minute ago:

- Added multi-language support
- Applied all recommend changes

**Comment by supechicken** [comment]
Update:
- Add crouton support
- Correct syntax

Tested on crouton `bullseye` chroot

File: packages/launcher.rb [hunk 35]
```diff
@@ -0,0 +1,27 @@
...
+    
+  def self.install
+    FileUtils.mkdir_p(CREW_DEST_PREFIX + '/bin')
+    FileUtils.install('../launcher.rb', CREW_DEST_PREFIX + '/bin/launcher', mode: 0755)
+  end
+    
+  def self.postinstall
+    puts 'Performing code-injection...'
+    @code = 'launcher start'
+    system "echo '#{@code}' >> #{HOME}/.bashrc" unless `grep -c '#{@code}' #{HOME}/.bashrc`.to_i > 0
```
Comment (uberhacker) [inline packages/launcher.rb:22]: Should we have a message to source ~/.bashrc?
```
puts "To finish the installation, execute the following:".lightblue
puts "source ~/.bashrc".lightblue
```

File: launcher/launcher.rb [hunk 36]
```diff
@@ -0,0 +1,298 @@
+#!/usr/bin/env ruby
+
+# use chromebrew variable if exist
+if File.exist?('/usr/local/lib/crew/lib/const.rb')
+  require_relative '/usr/local/lib/crew/lib/const'
```
Comment (uberhacker) [inline launcher/launcher.rb:5]: We should use `#{CREW_PREFIX}` instead of `/usr/local`.

File: launcher/launcher.rb (hunk 23 above)
Comment (uberhacker) [reply launcher/launcher.rb:284]: If the server failed to start, how would the user know?

**Review by uberhacker — CHANGES_REQUESTED** [review]

File: packages/launcher.rb (hunk 35 above)
Comment (supechicken) [reply packages/launcher.rb:22]: Added

**Comment by uberhacker** [comment]
@supechicken: Did you give up on this or are you working on an improved version?

**Comment by supechicken** [comment]
> @supechicken: Did you give up on this or are you working on an improved version?

I am working on PWA again with ruby built-in `socket` function since Chrome Apps are [deprecating now](https://blog.chromium.org/2020/01/moving-forward-from-chrome-apps.html) :) 

By the way can I open a repo in `github.com/chromebrew` for this?

**Comment by uberhacker** [comment]
Sure, if that helps you to organize the project better.

**Comment by daemonspudguy** [comment]
@supechicken is the repository for this project private?

**Comment by supechicken** [comment]
> @supechicken is the repository for this project private?

The repo was deleted by me, I will recreate it when the new PWA version is done

**Comment by brunnels** [comment]
Bump?  Just saw this is breaking ability to run flatpak apps.  The remount gets it working.

**Comment by daemonspudguy** [comment]
@supechicken I saw that the repo for the PWA version exists, but there is no package anywhere in your Chromebrew fork? Why?

**Comment by supechicken** [comment]
> @supechicken I saw that the repo for the PWA version exists, but there is no package anywhere in your Chromebrew fork? Why?

See supechicken/crew-launcher#1
